### DISCUSSION TITLE
Matrix approach: how to improve it using FP?

### DISCUSSION THREAD

**Discussion by Illegit** [freeplane/freeplane#851] [Ideas]
Hi, I'm looking for ideas on how to apply a matrix/tensor approach to place nodes and branches in each matrix element.
In my work I need to cluster at least 2 dimensions (sometimes more) of the following topics:

- solutions
- teams
- tasks
- performances

The matrix shape visualization is very helpful to understand the coverage of your approach to managing complex systems, and applying it in FP would be a strong enhancer of the managerial support it already provides to me.

Currently, I think of a branch of categories for each matrix dimension and a strong use of clones, but I find it difficult to apply the combining approach typical of matrixes. (it implies even a sort of visualization, that at now is not fully available in FP)
Any suggestion?
Thanks

**Comment by euu2021** [comment]
I'm not sure if I understand your needs. Are looking for something like relational databases? For example:

- Solution A implies tasks 1,2 and 3
- Team X has Solution A assigned to it
- So, if the user asks for the tasks assigned to Team X, the program automatically shows tasks 1,2 and 3

Do you want this, or something else?

**Comment by EdoFro** [comment]
Hi @Illegit,
I created a sample map as proof of concept to easily assign nodes to categories (using attributes) and to automatically group them.

Take a look at the map. It has three steps.

[Testing multiple dimensions classifications.zip](https://github.com/freeplane/freeplane/files/10124853/Testing.multiple.dimensions.classifications.zip)


It has also some sample categories created and nodes assigned to their different possible values. To change a value from a node or add a nother category to the node just move it to another branch.

It works using formulas, so it's not ready for 'production'. It's just to test an idea.

Best regards,

edo

**Comment by Illegit** [comment]
Thank you all for answering my question. 
@euu2021 I'll try to show better in 2 days what I mean, I need to draw a scheme.
@EdoFro : very interesting, maybe it fits only partially this topic but I think it has already use for it, I'll try in 2 days and I feedback (I already use your swap-o-matic, but this way it seems even smoother)

**Comment by Illegit** [comment]
@EdoFro: is it possible to apply your way in 2 different maps? because using clones is obliging to use 1 single map, and maybe using links to nodes can be solved, but still remains the formulas problem, what do you suggest?

**Reply by EdoFro** [reply]
I don't know. You have to explain your use case better.
What problem are you trying to solve?

**Comment by Illegit** [comment]
@EdoFro: your script is really interesting, is it possible in your script to add some characteristics? (sorry for my English)
1) sort the branch of results in ("node with icon 3") based on the value of the attributes (in decreasing or in increasing order of the values of the attributes)
2) change the date format in the attribute with dates (possibly to have the option to have this format "2022-12-23")
3) keep the format of each attribute value even in the branch of the results.

Thank you

**Reply by EdoFro** [reply]
done. take a look:

[Testing multiple dimensions classifications v2.zip](https://github.com/freeplane/freeplane/files/10157456/Testing.multiple.dimensions.classifications.v2.zip)

**Comment by Illegit** [comment]
Thanks it is ok. I applied it.
Some notes as reminders:
Risk if you just move nodes into 3 and you delete attributes in 3 you will
delete alla nodes, to protect copy or clone.
The formula a Little bit too heavy in memory usage eg for 150 nodes alla
cloned with 4 gb dedicate to FP it goes.
There is a format of text Like this: good%20morning, how to go with normale
text?
At now I reshaped the attributes values not to ha e spaces.

Il lun 5 dic 2022, 20:54 EdoFro ***@***.***> ha scritto:

**Reply by EdoFro** [reply]
hi @Illegit,
the attached mindmap with the formulas was just a test, something to try as proof of concept. Something to discuss about and for others to look at the scripts and get some new ideas.

Formulas shouldn't be used to modify nodes and branches. They are intended to just make some calcullation and reflect its result in its node.

For me, using them was an easy way to show how a mindmap could work if some tasks were automatized, for example with node change listeners, but it wasn't intended to work in big maps with many nodes in those branches.

ok?

**Reply by EdoFro** [reply]
> The formula a Little bit too heavy in memory usage eg for 150 nodes alla cloned with 4 gb dedicate to FP it goes.

that's because every little change makes the formula to recalculate and it also changes the map too.

I think I will make a new version of the map whith the scripts as attributes, and the user will have to execute them every time he wants to update the map.
I think that is good enough

**Reply by EdoFro** [reply]
> Risk if you just move nodes into 3 and you delete attributes in 3 you will delete alla nodes, to protect copy or clone.

For those cases I use custom styles or icons or attributes to diffentiate types of nodes. That way the script can easily recognize which nodes to delete.
In the new version I will include this.
What do you prefer: custom styles, icons or attributes to recognize the nodes that group the task nodes (in branch with the number 3 icon)?

**Reply by EdoFro** [reply]
> There is a format of text Like this: good%20morning, how to go with normale text?

I don't know why it thinks that the string values in attributes may be urls.
That's something that must be reviewed in detail.

I won't include this in a next version. Maybe somebody else can solve this issue.

If in the future this scripts are further used, then we can consider to transform this into an add on and solve this kindof problems.

ok?

**Reply by Illegit** [reply]
> hi @Illegit, the attached mindmap with the formulas was just a test, something to try as proof of concept. Something to discuss about and for others to look at the scripts and get some new ideas.
> 
> Formulas shouldn't be used to modify nodes and branches. They are intended to just make some calcullation and reflect its result in its node.
> 
> For me, using them was an easy way to show how a mindmap could work if some tasks were automatized, for example with node change listeners, but it wasn't intended to work in big maps with many nodes in those branches.
> 
> ok?

OK, I find it very helpful.

**Reply by Illegit** [reply]
> I think I will make a new version of the map whith the scripts as attributes, and the user will have to execute them every time he wants to update the map.
> I think that is good enough

Yes, I like it. Thank you.

**Reply by Illegit** [reply]
> What do you prefer: custom styles, icons or attributes to recognize the nodes that group the task nodes (in branch with the number 3 icon)?

I prefer attributes. Nice

**Reply by Illegit** [reply]
> If in the future this scripts are further used, then we can consider to transform this into an add on and solve this kindof problems.
> 
> ok?

It will be a good solution

**Comment by EdoFro** [comment]
> >Risk if you just move nodes into 3 and you delete attributes in 3 you will delete alla nodes, to protect copy or clone. The formula a Little bit too heavy in memory usage eg for 150 nodes alla cloned with 4 gb dedicate to FP it goes.

> I think I will make a new version of the map whith the scripts as attributes, and the user will have to execute them every time he wants to update the map.

Here is the new version:

[Testing multiple dimensions classifications v3.zip](https://github.com/freeplane/freeplane/files/10180092/Testing.multiple.dimensions.classifications.v3.zip)

**Reply by Illegit** [reply]
Hi EdoFRo, just to inform you: I mixed your 2 solutions formulas for and 2 with the script for 3, and it works very well. Automatical formulas to create the attributes and script to create the big branch with all of the clones. Nice.
I am already using it on some maps. Thank you. Probably it deserves an addon.

**Reply by Illegit** [reply]
Hi EdoFro, here is the resulting map I normally use following your solution. Thanks
[Matrix example.zip](https://github.com/freeplane/freeplane/files/10414939/Matrix.example.zip)

**Comment by Illegit** [comment]
Hi EdoFro, is it possible to change the AOM writing in the details of each node generated by your script1 (in the node number 3)  in a formula like the following =children.size() ? 
Thanks

**Comment by Illegit** [comment]
Hi EdoFro, I have to correct my last comment: when I run the script to create the visualization (node number 3 in your file V3) I see the name of the attribute in the details of the node, I like to add to that description in details even the following formula: =children.size(), how to do it? Thanks

**Reply by EdoFro** [reply]
here is a new version:
[Testing multiple dimensions classifications v4.zip](https://github.com/freeplane/freeplane/files/10354878/Testing.multiple.dimensions.classifications.v4.zip)

hope it helps

**Comment by Illegit** [comment]
Hi EdoFro, thank you for your script, but It creates to me some strange notes (see point 1)
![Immagine 2023-01-06 112436](https://user-images.githubusercontent.com/62674157/210986052-b528798c-15f0-4535-b93d-c8a73f26f214.png)
and it cancels the attribute identification and writes only the children's size,
here instead what I mean, I created a very simple script for adding the formula to the current details, but I do not know how to insert in your script node3 version V3.
![Immagine 2023-01-06 112107](https://user-images.githubusercontent.com/62674157/210984584-88335789-4adb-4a8c-9a9e-c68177ed9170.png)
Thank you even at this level version V3 I find it very useful.

**Comment by Illegit** [comment]
here the simple script activated with a key
[DetailsAndSomething.zip](https://github.com/freeplane/freeplane/files/10359387/DetailsAndSomething.zip)

**Comment by Illegit** [comment]
[DetailsAndSomething.zip](https://github.com/freeplane/freeplane/files/10359468/DetailsAndSomething.zip)
This one is better
